### PR TITLE
tidb-lightning: document max-error and duplicate-resolution

### PR DESCRIPTION
**Pull request by hfxsd**
### What is changed, added or deleted? (Required)

Document the error resolution features, introduced in pingcap/tidb#27181 and pingcap/tidb#28464.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [x] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/8132
- Other reference link(s):https://github.com/pingcap/docs/pull/6717

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by gozssky — APPROVED** [review]

**Review by glorv — COMMENTED** [review]
LGTM

**Review by gozssky — APPROVED** [review]

File: tidb-lightning/tidb-lightning-configuration.md [hunk 1]
```diff
@@ -63,6 +63,15 @@ table-concurrency = 6
 # medium, this value might need to be adjusted for optimal performance.
 io-concurrency = 5
 
+# maximum number of non-fatal errors to tolerate before stopping Lightning.
+# Non-fatal errors are those that are localized to a few rows, and ignoring those rows allows the import process to continue.
+# Setting this to N means that Lightning will stop as soon as possible when the (N+1)-th error is encountered.
+# The skipped rows will be inserted to tables inside the "task info" schema on the target TiDB, which can be configured below.
+max-error = 0
+# task-info-schema-name is the name of the schema/database storing human-readable Lightning execution result.
```
Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-configuration.md:71]: I think it's not only human-readable but also for programs.

**Review by sunzhaoyang — APPROVED** [review]
rest LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+* Invalid value (example: set `'Text'` to an INT column).
+* Numeric overflow (example: set 500 to a TINYINT column)
+* String overflow (example: set `'Very Long Text'` to a VARCHAR(5) column).
+* Zero date time (namely `'0000-00-00'` and `'2021-12-00'`).
+* Set NULL to a NOT NULL column.
+* Failed to evaluate a generated column expression.
+* Column count mismatch. Number of values in the row does not match number of columns of the table.
+* Unique/Primary key conflict in TiDB backend, when `on-duplicate = "error"`.
+* Any other SQL errors.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:39]:
```suggestion
* Invalid values (example: set `'Text'` to an INT column).
* Numeric overflow (example: set `500` to a TINYINT column)
* String overflow (example: set `'Very Long Text'` to a VARCHAR(5) column).
* Zero date-time (namely `'0000-00-00'` and `'2021-12-00'`).
* Set NULL to a NOT NULL column.
* Failed to evaluate a generated column expression.
* Column count mismatch. The number of values in the row does not match the number of columns of the table.
* Unique/Primary key conflict in TiDB-backend, when `on-duplicate = "error"`.
* Any other SQL errors.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+The following errors are always fatal, and cannot be skipped by changing `max-error`:
+
+* Syntax error (such as unclosed quotation marks) in the original CSV, SQL or Parquet file.
+* I/O, network or system permission errors.
+
+Unique/Primary key conflict in Local backend is handled separately, explained in the next section.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:46]:
```suggestion
Unique/Primary key conflict in the Local-backend is handled separately and explained in the next section.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+## Duplicate resolution in Local-backend mode
+
+Local-backend mode imports data by first converting them to KV pairs, and ingest them into TiKV in batches. Unlike the TiDB-backend mode, duplicate rows are not detected until the end of task. Therefore, duplicate errors in Local-backend mode is not controlled by `max-error`, but rather a separate configuration `duplicate-resolution`.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:50]:
```suggestion
In the Local-backend mode, TiDB Lightning imports data by first converting them to KV pairs and ingesting the pairs into TiKV in batches. Unlike the TiDB-backend mode, duplicate rows are not detected until the end of a task. Therefore, duplicate errors in the Local-backend mode are not controlled by `max-error`, but rather by a separate configuration `duplicate-resolution`.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+{{< copyable "" >}}
+
+```toml
+[tikv-importer]
+duplicate-resolution = 'record'
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:56]:
```suggestion
duplicate-resolution = 'none'
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+```
+
+There are three possible values of `duplicate-resolution`:
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:59]:
```suggestion
The value options of `duplicate-resolution` are as follows:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+* **'none'** — Do not detect duplicates. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
+* **'record'** — Detect duplicates, but do not attempt to fix it. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
+* **'remove'** — Detect duplicates, and remove *all* duplicated rows. The imported table will be consistent, but the involved rows are ignored and have to be added back manually.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:63]:
```suggestion
* **'none'**: Does not detect duplicate data. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
* **'record'**: Detects duplicate data, but does not attempt to fix it. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
* **'remove'**: Detects duplicate data, and removes *all* duplicated rows. The imported table will be consistent, but the involved rows are ignored and have to be added back manually.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+TiDB Lightning duplicate resolution can only detect duplicates within the data source. It cannot handle conflict with existing data before running TiDB Lightning.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:65]:
```suggestion
TiDB Lightning duplicate resolution can detect duplicate data only within the data source. This feature cannot handle conflict with existing data before running TiDB Lightning.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+You can change the database name by configuring `lightning.task-info-schema-name`.
+
+{{< copyable "" >}}
+
+```toml
+[lightning]
+task-info-schema-name = 'lightning_task_info'
+```
+
+TiDB Lightning creates 3 tables inside this database:
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:80]:
```suggestion
TiDB Lightning creates 3 tables in this database:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+    row_data    text NOT NULL,
+    raw_key     mediumblob NOT NULL,
+    raw_value   mediumblob NOT NULL,
+    raw_handle  mediumblob NOT NULL,
+    raw_row     mediumblob NOT NULL,
+    KEY (task_id, table_name)
+);
+```
+
+**syntax_error_v1** is intended to record syntax error from files. It is not implemented yet.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:118]: 未生效的项要写入文档吗？

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+
+**type_error_v1** records all [type errors](#type-error) managed by the `max-error` configuration. There is one row per error.
+
+**conflict_error_v1** records all [unique/primary key conflict in local backend](#duplicate-resolution). There are 2 rows per pair of conflict.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:122]:
```suggestion
**conflict_error_v1** records all [unique/primary key conflict in the Local-backend](#duplicate-resolution). There are 2 rows per pair of conflicts.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+| Column       | syntax | type | conflict | Explanation                                                                                                                         |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:124]:
```suggestion
| Column       | Syntax | Type | Conflict | Description                                                                                                                         |
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+| ------------ | ------ | ---- | -------- | ----------------------------------------------------------------------------------------------------------------------------------- |
+| task_id      | ✓      | ✓    | ✓        | The TiDB Lightning task ID generating this error                                                                                    |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:126]:
```suggestion
| task_id      | ✓      | ✓    | ✓        | The TiDB Lightning task ID that generates this error                                                                                    |
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+| create_table | ✓      | ✓    | ✓        | When the error was recorded                                                                                                         |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:127]:
```suggestion
| create_table | ✓      | ✓    | ✓        | The time at which the error is recorded                                                                                                         |
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+| table_name   | ✓      | ✓    | ✓        | Name of the table containing the error, in the form ``'`db`.`tbl`'``                                                                |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:128]:
```suggestion
| table_name   | ✓      | ✓    | ✓        | The name of the table that contains the error, in the form of ``'`db`.`tbl`'``                                                                |
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+| key_data     |        |      | ✓        | Formatted key handle of the row causing the error. The content is for human reference only, and not intended to be machine readable |
+| row_data     |        | ✓    | ✓        | Formatted row data causing the error. The content is for human reference only, and not intended to be machine readable              |
+| raw_key      |        |      | ✓        | Key of the conflicted KV pair                                                                                                       |
+| raw_value    |        |      | ✓        | Value of the conflicted KV pair                                                                                                     |
+| raw_handle   |        |      | ✓        | Row handle of the conflicted row                                                                                                    |
+| raw_row      |        |      | ✓        | Encoded value of the conflicted row                                                                                                 |
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:139]:
```suggestion
| path         | ✓      | ✓    |          | The path of the file that contains the error                                                                                               |
| offset       | ✓      | ✓    |          | The byte position in the file where the error is found                                                                                  |
| error        | ✓      | ✓    |          | The error message                                                                                                                       |
| context      | ✓      |      |          | The text that surrounds the error                                                                                                          |
| index_name   |        |      | ✓        | The name of the unique key in conflict. It is `'PRIMARY'` for primary key conflicts.                                                          |
| key_data     |        |      | ✓        | The formatted key handle of the row that causes the error. The content is for human reference only, and not intended to be machine-readable. |
| row_data     |        | ✓    | ✓        | The formatted row data that causes the error. The content is for human reference only, and not intended to be machine-readable              |
| raw_key      |        |      | ✓        | The key of the conflicted KV pair                                                                                                       |
| raw_value    |        |      | ✓        | The value of the conflicted KV pair                                                                                                     |
| raw_handle   |        |      | ✓        | The row handle of the conflicted row                                                                                                    |
| raw_row      |        |      | ✓        | The encoded value of the conflicted row                                                                                                 |
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+
+> **Note:**
+>
+> The error report records the file offset, not line/column number which is inefficient to obtain. You can quickly jump near a byte position (using 183 as example) with
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:143]:
```suggestion
> The error report records the file offset, not line/column number which is inefficient to obtain. You can quickly jump near a byte position (using 183 as example) using the following commands:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+>
+> * shell, printing several lines before —
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:145]:
```suggestion
> * shell, printing the first several lines.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+>
+>     ```shell
+>     head -c 183 file.csv | tail
+>     ```
+>
+> * shell, printing several lines after —
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:151]:
```suggestion
> * shell, printing the next several lines:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+>
+>     ```shell
+>     tail -c +183 file.csv | head
+>     ```
+>
+> * vim — `:goto 183` or `183go`
+
+## Example
+
+In this example we prepare a data source with some known errors.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:161]:
```suggestion
In this example, a data source is prepared with some known errors.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+
+1. Prepare the database and table schemas.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:163]:
```suggestion
1. Prepare the database and table schema.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+        (54, 'fifty-four'),     -- conflicts with the other 'fifty-four' below
+        (77, 'seventy-seven'),  -- the string is longer than 12 characters
+        (600, 'six hundred'),   -- the number overflows TINYINT
+        (40, 'fourty'),         -- conflicts with the other 40 above
+        (42, 'fifty-four');     -- conflicts with the other 'fifty-four' above
+
+    EOF
+    ```
+
+3. Configure TiDB Lightning enable strict SQL mode, use the local backend to import, resolve duplicates by deleting, and skip up to 10 errors.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:195]:
```suggestion
3. Configure TiDB Lightning to enable strict SQL mode, use the Local-backend to import data, delete duplicates, and skip up to 10 errors.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+
+4. Run TiDB Lightning. This command will exit successfully because all errors are skipped.
+
+    {{< copyable "shell-regular" >}}
+
+    ```shell
+    tiup tidb-lightning -c config.toml
+    ```
+
+5. Verify that the imported table contains just the two normal rows:
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:230]:
```suggestion
5. Verify that the imported table only contains the two normal rows:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+    $ mysql -u root -h 127.0.0.1 -P 4000 -e 'select * from example.t'
+    +---+-----+
+    | a | b   |
+    +---+-----+
+    | 1 | one |
+    | 2 | two |
+    +---+-----+
+    ```
+
+6. Check that `type_error_v1` table caught the three rows involving type conversion:
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:242]:
```suggestion
6. Check whether the `type_error_v1` table has caught the three rows involving type conversion:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+        task_id: 1635888701843303564
+    create_time: 2021-11-02 21:31:42.629929
+     table_name: `example`.`t`
+           path: example.t.1.sql
+         offset: 253
+          error: failed to cast value as tinyint(4) for column `a` (#1): [types:1690]constant 600 overflows tinyint
+       row_data: (600,'six hundred')
+    ```
+
+7. Check that `conflict_error_v1` table caught the four rows having unique/primary key conflicts:
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:275]:
```suggestion
7. Check whether the `conflict_error_v1` table has caught the four rows that have unique/primary key conflicts:
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+
+```toml
+[tikv-importer]
+duplicate-resolution = 'record'
+```
+
+There are three possible values of `duplicate-resolution`:
+
+* **'none'** — Do not detect duplicates. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
+* **'record'** — Detect duplicates, but do not attempt to fix it. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
```
Comment (sunzhaoyang) [inline tidb-lightning/tidb-lightning-error-resolution.md:62]: @gozssky Please confirm this part

File: tidb-lightning/tidb-lightning-configuration.md [hunk 27]
```diff
@@ -63,6 +63,15 @@ table-concurrency = 6
...
+# set this to empty string to disable error recording.
+# task-info-schema-name = 'lightning_task_info'
```
[the remark this answers is not in the document]
Comment (hfxsd) [reply tidb-lightning/tidb-lightning-configuration.md:73]: 看起来是漏翻了。而且中文还有 meta-schema-name = "lightning_metadata"。我更新下。

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 26 above)
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:62]: `will fail checksum check` -> `will skip checksum and report conflict errors count`

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 5 above)
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:56]: LGTM

File: tidb-lightning/tidb-lightning-configuration.md [hunk 28]
```diff
@@ -107,6 +116,13 @@ addr = "172.16.31.10:8287"
 #  - ignore: keep the existing entry, and ignore the new entry
 #  - error: report error and quit the program
 # on-duplicate = "replace"
+# Whether to detect and resolve duplicate records (unique key conflict) when the backend is 'local'.
+# Current supports three resolution algorithms:
+#  - record: only records duplicate records to `lightning_task_info.conflict_error_v1` table on the target TiDB. Note that this
+#    required the version of target TiKV version is no less than v5.2.0, otherwise it will fallback to 'none'.
+#  - none: doesn't detect duplicate records, which has the best performance of the three algorithms, but probably leads to
+#    inconsistent data in the target TiDB.
+#  - remove: records all duplicate records like the 'record' algorithm and remove all duplicate records to ensure a consistent
```
[the remark this answers is not in the document]
Comment (gozssky) [reply tidb-lightning/tidb-lightning-configuration.md:125]: 这段描述确实容易引起歧义，准确来说是 record all duplicate records to task info database 和 remove all duplicate records from the target table。

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 29]
```diff
@@ -0,0 +1,327 @@
...
+
+```toml
+[tikv-importer]
+duplicate-resolution = 'record'
+```
+
+The value options of `duplicate-resolution` are as follows:
+
+* **'none'** — Do not detect duplicates. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
+* **'record'** — Detect duplicates, but do not attempt to fix it. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will fail checksum check.
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:62]:
```suggestion
* **'record'** — Detect duplicates, but do not attempt to fix it. If a unique/primary key conflict does exist, the imported table will have inconsistent data and index, and will skip checksum and report the count of the conflict error s.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 30]
```diff
@@ -0,0 +1,327 @@
...
+    row_data    text NOT NULL,
+    raw_key     mediumblob NOT NULL,
+    raw_value   mediumblob NOT NULL,
+    raw_handle  mediumblob NOT NULL,
+    raw_row     mediumblob NOT NULL,
+    KEY (task_id, table_name)
+);
+```
+
+**syntax_error_v1** is intended to record syntax error from files. It is not implemented yet.
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:118]:
```suggestion
/**
**syntax_error_v1** is intended to record syntax error from files. It is not implemented yet.
**/
```

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 10 above)
Comment (hfxsd) [reply tidb-lightning/tidb-lightning-error-resolution.md:118]: 先注释掉了

File: tidb-lightning/tidb-lightning-configuration.md (hunk 1 above)
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:71]:
```suggestion
# task-info-schema-name is the name of the schema/database storing Lightning execution results.
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 31]
```diff
@@ -0,0 +1,329 @@
...
+    key_data    text NOT NULL,
+    row_data    text NOT NULL,
+    raw_key     mediumblob NOT NULL,
+    raw_value   mediumblob NOT NULL,
+    raw_handle  mediumblob NOT NULL,
+    raw_row     mediumblob NOT NULL,
+    KEY (task_id, table_name)
+);
+```
+
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:117]:
```suggestion
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 32]
```diff
@@ -63,6 +63,15 @@ table-concurrency = 6
...
 io-concurrency = 5
 
+# maximum number of non-fatal errors to tolerate before stopping Lightning.
+# Non-fatal errors are those that are localized to a few rows, and ignoring those rows allows the import process to continue.
+# Setting this to N means that Lightning will stop as soon as possible when the (N+1)-th error is encountered.
+# The skipped rows will be inserted to tables inside the "task info" schema on the target TiDB, which can be configured below.
+max-error = 0
+# task-info-schema-name is the name of the schema/database storing Lightning execution results.
+# set this to empty string to disable error recording.
+# task-info-schema-name = 'lightning_task_info'
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-configuration.md:73]:
```suggestion
# The maximum number of non-fatal errors to tolerate before stopping TiDB Lightning.
# Non-fatal errors are localized to a few rows, and ignoring those rows allows the import process to continue.
# Setting this to N means that TiDB Lightning will stop as soon as possible when the (N+1)-th error is encountered.
# The skipped rows will be inserted into tables inside the "task info" schema on the target TiDB, which can be configured below.
max-error = 0
# task-info-schema-name is the name of the schema or database that stores TiDB Lightning execution results.
# To disable error recording, set this to an empty string.
# task-info-schema-name = 'lightning_task_info'
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 33]
```diff
@@ -120,6 +129,13 @@ addr = "172.16.31.10:8287"
 #  - ignore: keep the existing entry, and ignore the new entry
 #  - error: report error and quit the program
 # on-duplicate = "replace"
+# Whether to detect and resolve duplicate records (unique key conflict) when the backend is 'local'.
+# The following resolution algorithms are supported:
+#  - record: only records duplicate records to the `lightning_task_info.conflict_error_v1` table on the target TiDB. Note that the
+#    required version of the target TiKV is no earlier than v5.2.0; otherwise it falls back to 'none'.
+#  - none: does not detect duplicate records, which has the best performance of the three algorithms, but might lead to
+#    inconsistent data in the target TiDB.
+#  - remove: records all duplicate records to the lightning_task_info database, like the 'record' algorithm. But it removes all duplicate records from the target table to ensure a consistent.
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-configuration.md:138]:
```suggestion
#  - remove: records all duplicate records to the lightning_task_info database, like the 'record' algorithm. But it removes all duplicate records from the target table to ensure a consistent
```

**Review by TomShawn — APPROVED** [review]
Rest LGTM

**Comment by TomShawn** [comment]
LGTM

File: tidb-lightning/tidb-lightning-configuration.md [hunk 34]
```diff
@@ -63,6 +63,15 @@ table-concurrency = 6
...
 io-concurrency = 5
 
+# The maximum number of non-fatal errors to tolerate before stopping TiDB Lightning.
+# Non-fatal errors are localized to a few rows, and ignoring those rows allows the import process to continue.
+# Setting this to N means that TiDB Lightning will stop as soon as possible when the (N+1)-th error is encountered.
+# The skipped rows will be inserted into tables inside the "task info" schema on the target TiDB, which can be configured below.
+max-error = 0
+# task-info-schema-name is the name of the schema or database that stores TiDB Lightning execution results.
+# To disable error recording, set this to an empty string.
+# task-info-schema-name = 'lightning_task_info'
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-configuration.md:73]: Follow-up: To translate this part to Chinese

Comment (hfxsd) [reply tidb-lightning/tidb-lightning-configuration.md:73]: See PR: https://github.com/pingcap/docs-cn/pull/8211

**Pull request merged**